packages/runtime-core/src/apiOptions.ts

```diff
@@ -239,42 +241,81 @@ export function applyOptions(
     if (!isObject(data)) {
       __DEV__ && warn(`data() should return an object.`)
     } else if (instance.data === EMPTY_OBJ) {
-      instance.data = reactive(data)
+      let uniqueData: Record<any, any> = {}
```

Comment:
This while for in loop should be dev only to avoid the perf cost. It's ok for data to override props as long as there is a warning.